data_generators/xcspec_extractor.py

```diff
@@ -63,29 +65,117 @@ def xcspec_from_file(path):
     "com.apple.xcode.tools.swift.compiler",
 )
 
+cond_contents = """
+# foo
+
+"""
+
 bzl_contents = """
 # Extracted from Xcode %s
 # To update, in rules_ios run `bazel run data_generators:extract_xcspecs`
 
 SETTINGS = """ % xcode_version
 
+meths = {}
+
+def str_to_thing(string, key):
+    r = re.compile("\$\(([^\$)]+)\)")
+    if not r.search(string):
+        return 'return (False, %s)' % repr(string)
+
+    d = {}
+
+    def record_substitution(match):
+        name = match[1]
+        if name not in d:
+            d[name] = (
+                'eval_val_%s' % len(d),
+                name,
+                'eval_key_%s' % len(d),
+            )
+        if key == "DefaultValue":
+            return "{%s}" % d[name][0]
+        return d[name][0]
+
+    while True:
+        (string, subs_made) = r.subn(repl=record_substitution, string=string)
+        if not subs_made:
+            break
+
+    def format_str(s):
+        f = re.findall('{(eval_val_\d+)}', s)
+        if not f:
+            return '"%s"' % s
+        if len(f) == 1:
+            if s == '{%s}' % f[0]:
+                return f[0]
+        return '"%s".format(%s)' % (s, ', '.join(('{var} = {var}'.format(var = v) for v in f)))
+
+    s = "\n    used_user_content = False\n" + "\n".join((
+        """
+    {var} = ""
+    {key} = {key_value}
+    if {key} in xcconfigs:
+        {var} = xcconfigs[{key}]
+        used_user_content = True
+    elif {key} in id_configs:
+        opt = id_configs[{key}]
+        if "DefaultValue" in opt:
+            ({var}_used_user_content, {var}) = XCSPEC_EVALS[opt["DefaultValue"]](xcconfigs, id_configs)
+            used_user_content = used_user_content or {var}_used_user_content""".format(var=var, name=name, key=key, key_value=format_str(name))
+        for (name, (var, _i, key)) in d.items()
+    ))
+
+    if key == 'DefaultValue':
+        return s + '\n\n    return (used_user_content, %s)' % (format_str(string), )
+    elif key == 'Condition':
+        string = re.sub(pattern="(?:^|[(&|!= ])(?!eval_val_\d+)([A-Za-z0-9_]+)", repl= lambda m: repr(m[1]), string=string)
+        return s + '\n\n    return (used_user_content, (%s))' % string.replace('||', 'or').replace('!=', '__NEQ__').replace('!', 'not ').replace('&&', 'and').replace('__NEQ__', '!=')
+    
+    raise "Unknown key %s" % key
+
+def _add_eval(id: str, name, key, string, meths):
+    meth = '__'.join((id.replace('.', '_'), name.replace('.', '_'), key))
```

Comment:
maybe mangle more, so search results don't get cluttered when grepping for a particular xcconfig name?